trailmet/algorithms/quantize/quantize.py

```diff
@@ -0,0 +1,181 @@
+
+import torch
+import torch.nn as nn
+import torch.nn.init as init
+from tqdm import tqdm_notebook
+from ..algorithms import BaseAlgorithm
+
+class BaseQuantization(BaseAlgorithm):
+    """base class for quantization algorithms"""
+    def __init__(self, **kwargs):
+        super(BaseQuantization, self).__init__(**kwargs)
+        pass
+
+    def quantize(self, model, dataloaders, method, **kwargs):
+        pass
+    
+    def round_ste(x: torch.Tensor):
+        """
+        Implement Straight-Through Estimator for rounding operation.
+        """
+        return (x.round() - x).detach() + x
+
+    def get_calib_samples(self, train_loader, num_samples):
+        """
+        Get calibration-set samples for finetuning weights and clipping parameters
+        """
+        calib_data = []
+        for batch in train_loader:
+            calib_data.append(batch[0])
+            if len(calib_data)*batch[0].size(0) >= num_samples:
+                break
+        return torch.cat(calib_data, dim=0)[:num_samples]
+
+# the lp_loss, accuracy and test functions can be shifted to the BaseAlogorithm class in ./algorithms.py
+    
+    def lp_loss(pred, tgt, p=2.0, reduction='none'):
+        """loss function measured in Lp Norm"""
+        if reduction == 'none':
+            return (pred-tgt).abs().pow(p).sum(1).mean()
+        else:
+            return (pred-tgt).abs().pow(p).mean()
+
+    def accuracy(self, output, target, topk=(1,)):
+        """Computes the accuracy over the k top predictions for the specified values of k"""
+        with torch.no_grad():
+            maxk = max(topk)
+            batch_size = target.size(0)
+
+            _, pred = output.topk(maxk, 1, True, True)
+            pred = pred.t()
+            correct = pred.eq(target.view(1, -1).expand_as(pred))
+
+            res = []
+            for k in topk:
+                correct_k = correct[:k].reshape(-1).float().sum(0, keepdim=True)
+                res.append(correct_k.mul_(100.0 / batch_size))
+            return res
+
+    def test(self, model, dataloader, loss_fn=None, device=None):
```

Comment:
this is already implemented, can we combine the both?

Reply:
done in e52f86293114e05d095189e26f3c0dcce71d6235